hw/ip/kmac/doc/_index.md

```diff
@@ -0,0 +1,274 @@
+
+The software completes the operation by issuing Done command after reading the digest.
+THe padding logic clears internal variables and goes back to Idle state.
+
+### Padding for KMAC
+
+![](kmac-padding.svg)
+
+KMAC core prepends and appends additional bitstream on top of Keccak padding logic in SHA3 core.
+The [NIST SP 800-185][] defines `KMAC[128,256](K, X, L, S)` as a cSHAKE function.
+See the section 4.3 in NIST SP 800-185 for details.
+If KMAC is enabled, the software should configure !!CFG.mode to cSHAKE and the first six bytes of !!PREFIX to `0x01204B4D4143` (bigendian).
+The first six bytes of !!PREFIX represents the value of `encode_string("KMAC")`.
+
+The KMAC padding logic prepends a block containing the encoded secret key to the output message.
+The KMAC first sends the block of secret key then accepts the incoming message bitstream.
+At the end of the message, the software writes `right_encode(output_length)` to MSG_FIFO prior to issue Process command.
+
+### Message FIFO
+
+The KMAC HWIP has a compile-time configurable depth message FIFO inside.
+The message FIFO receives incoming message bitstream regardless of its byte position in a word.
+Then it packs the partial message bytes into the internal 64 bit data width.
+After packing the data, the logic stores the data into the FIFO until the internal KMAC/SHA3 engine consumes the data.
+
+The depth of the message FIFO is chosen to cover the throughput of the software or other producers such as DMA engine.
+The size of the message FIFO is enough to hold the incoming data while the SHA3 engine is processing the previous block.
+Details are in `kmac_pkg::MsgFifoDepth` parameter.
+Default design parameters assume the system characteristics as below:
+
+- `kmac_pkg::RegLatency`: The register write takes 5 cycles.
+- `kmac_pkg::Sha3Latency`: Keccak round latency takes 72 cycles, which is the masked version of the Keccak round.
+
+The message FIFO does not generate the masked message data.
+Incoming message bistream is not sensitive to the leakage.
+So if the `EnMasking` parameter is set, the second share of the message bistream is zero.
+The secret key, however, is stored as masked form always.
+
+### Keccak State Access
+
+After the Keccak round completes the KMAC/SHA3 operation, the contents of the Keccak state contain the digest value.
+The software can access the 1600 bit of the Keccak state directly through the window of the KMAC/SHA3 register.
+
+If the compile-time parameter masking feature is enabled, the upper 256B of the window is the second share of the Keccak state.
+If not, the upper address space is zero value.
+The software reads both of the Keccak state shares and XORed in the software to get the unmasked digest value if masking feature is set.
+
+The Keccak state is valid after the Sponge absorbing process is completed.
```

Comment:
very small nit: does `Sponge` need to be capitalized here? seems like it's not capitalized in the next sentence